TV Casting API docs (#178)
sharadb-amazon/connectedhomeip: examples/tv-casting-app/APIs.md

[iOS](darwin) to understand how to build and consume each platform's specicific
libraries.

### Initialize the Casting Client

_{Complete Initialization examples:
[Android](android/App/app/src/main/java/com/matter/casting/InitializationExample.java)
| [iOS](darwin/TvCasting/TvCasting/TvCastingApp.swift)}_

A Casting Client must first initialize the Matter SDK and define the following
[`DataProvider`](android/App/app/src/main/jni/com/matter/casting/params/DataProvider.java)
objects for the the Matter Casting library to use throughout the client's
lifecycle:

1.  **Rotating Device Identifier** - Refer to the Matter specification for
    details on how to generate the
    [Rotating Device Identifier](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/rendezvous/DeviceDiscovery.adoc#245-rotating-device-identifier)).
    Then, instantiate a `DataProvider` object as described below.

    On Android, define a `rotatingDeviceIdUniqueIdProvider` to provide the
    Casting Client's RotatingDeviceIdUniqueId, by implementing a `DataSource`:

    ```java
    private final static DataProvider<byte[]> rotatingDeviceIdUniqueIdProvider = new DataProvider<byte[]>() {

>         private static final String APP_ID = "YOUR_GENERATED_ROTATING_DEVICE_ID";

Reviewer: is APP_ID the best name for this? Also, shouldn't a real implementation pass the value in as an initializer argument given that it will likely be generated by the cloud or a random number generator?

Author: I've renamed it to ROTATING\_DEVICE\_ID\_UNIQUE\_ID to make what it is more obvious. 

About the initializer arg question - the DataProvider that provides this value is itself being passed as an initializer argument to the SDK.

Author: Should I add a comment saying that these are just sample values and you can pass any valid value to the SDK?

Author: Done

        @Override
        public byte[] get() {
            return APP_ID.getBytes();
        }
    }
    ```

    On iOS, define the
    `func castingAppDidReceiveRequestForRotatingDeviceIdUniqueId` in a class,
    `AppParametersDataSource`, that implements the `MTRDataSource`:

    ```objectivec
    class AppParametersDataSource : NSObject, MTRDataSource
    {
        func castingAppDidReceiveRequestForRotatingDeviceIdUniqueId(_ sender: Any) -> Data {
            return "YOUR_GENERATED_ROTATING_DEVICE_ID".data(using: .utf8)!
        }
        ...
    }
    ```

2.  **Commissioning Data** - This object contains the passcode, discriminator,

>     etc needed to verify an incoming commissioning request. Refer to the Matter

Reviewer: ```suggestion
    etc which identify the app and are provided to the CastingPlayer during the commissioning process. Refer to the Matter
```

    specification's
    [Onboarding Payload](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/qr_code/OnboardingPayload.adoc#ref_OnboardingPayload))
    section for details on commissioning data.

    On Android, define a `commissioningDataProvider` that can provide the
    required values to the CastingApp.

    ```java
    private final static DataProvider<CommissioningData> commissioningDataProvider = () -> {
        CommissioningData commissioningData = new CommissioningData();

>         commissioningData.setSetupPasscode(20202021);

Reviewer: shouldn't the passcode and discriminator be passed in as arguments (the dummy values could be included as comments) since they should not be hard coded?

Author: The DataProvider itself is passed in as an argument. The values it is providing currently are just sample values that would work with the example tv-app (since the tv-app hardcodes these values as well).

Reviewer 2: maybe its useful to add a comment indicating these are dummy values or something

    `AppParametersDataSource` class defined above, that can provide the required
    values to the `MTRCastingApp`.

    ```objectivec
    func castingAppDidReceiveRequestForCommissioningData(_ sender: Any) -> MTRCommissioningData {
        return MTRCommissioningData(
            passcode: 20202021,
            discriminator: 3874,
            spake2pIterationCount: 1000,
            spake2pVerifier: nil,
            spake2pSalt: nil)
    }
    ```

3.  **Device Attestation Credentials** - This object contains the
    `DeviceAttestationCertificate`, `ProductAttestationIntermediateCertificate`,
    etc. and implements a way to sign messages when called upon by the Matter TV
    Casting Library as part of the
    [Device Attestation process](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/device_attestation/Device_Attestation_Specification.adoc)
    during commissoning.

    On Android, define a `dacProvider` to provide the Casting Client's Device
    Attestation Credentials, by implementing a `DataSource`:

    ```java
    private final static DataProvider<DeviceAttestationCredentials> dacProvider = new DataProvider<DeviceAttestationCredentials>() {

>         private static final String kDevelopmentDAC_Cert_FFF1_8001 = "MIIB5z...<snipped>...CXE1M=";

Reviewer: Seems weird to hard code these values in the sample code.

Author: Same answer as above.

implements the `CastingPlayerChangeListener`.

```java
private static final CastingPlayerChangeListener castingPlayerChangeListener = new CastingPlayerChangeListener()
{
    private final String TAG = CastingPlayerChangeListener.class.getSimpleName();

    @Override
    public void onChanged(List<CastingPlayer> castingPlayers) {
        Log.i(TAG, "Discovered changes to " + castingPlayers.size() + " CastingPlayers");
        // consume changes to the provided castingPlayers
    }

    @Override
    public void onAdded(List<CastingPlayer> castingPlayers) {
        Log.i(TAG, "Discovered " + castingPlayers.size() + " CastingPlayers");
        // consume discovered castingPlayers
    }

    @Override
    public void onRemoved(List<CastingPlayer> castingPlayers) {
        Log.i(TAG, "Removed " + castingPlayers.size() + " CastingPlayers");
        // consume castingPlayers removed or lost from the network
    }
};
```

On iOS, implement a `func addDiscoveredCastingPlayers`,
`func removeDiscoveredCastingPlayers` and `func updateDiscoveredCastingPlayers`
which listen to notifications as Casting Players are added, removed, or updated.

```objectivec
@objc

> func addDiscoveredCastingPlayers(notification: Notification)

Reviewer 3: The function names do not really align with them being a handle of a notification. Maybe something like didDiscoverCastingPlayers? The same applies for the other functions.

Author: Done

    // After some time, stop discovering and remove our castingPlayerChangeListener
    // from the set of listeners the CastingPlayerDiscoverer informs
    executorService.schedule(() -> {
        discoverer.stopDiscovery();
        discoverer.removeCastingPlayerChangeListener(castingPlayerChangeListener);
    },     30, TimeUnit.SECONDS);
}
```

On iOS, register the listeners by calling `addObserver` on the
`NotificationCenter` with the appropriate selector, and then call `start` on the
`sharedInstance` of `MTRCastingPlayerDiscovery`.

```objectivec
func startDiscovery() {
    NotificationCenter.default.addObserver(
    self,
    selector: #selector(self.addDiscoveredCastingPlayers),
    name: NSNotification.Name.didAddCastingPlayers,
    object: nil)
    NotificationCenter.default.addObserver(
    self,
    selector: #selector(self.removeDiscoveredCastingPlayers),
    name: NSNotification.Name.didRemoveCastingPlayers,
    object: nil)
    NotificationCenter.default.addObserver(
    self,
    selector: #selector(self.updateDiscoveredCastingPlayers),
    name: NSNotification.Name.didUpdateCastingPlayers,
    object: nil)

    MTRCastingPlayerDiscovery.sharedInstance().start()

>     self.discoveryRequestStatus = true

Reviewer 3: What is discoveryRequestStatus needed for?

Author: This was just a state variable I waqs using to update a UI element on the sample app. I have replaced it with an "..." in the documentation to avoid confusing the reader.

}
```

Note: You will need to connect with a Casting Player as described below to see
the list of Endpoints that they support. Refer to the [Connection](#connection)
section for details on how to discover available endpoints supported by a
Casting Player.

### Connect to a Casting Player

_{Complete Connection examples:
[Android](android/App/app/src/main/java/com/matter/casting/ConnectionExample.java)
| [iOS](darwin/TvCasting/TvCasting/MTRConnectionExampleViewModel.swift)}_

Each CastingPlayer object created during [Discovery](#discovery) contains
information such as deviceName, vendorId, productId, etc. which can help the
user pick the right CastingPlayer. A Casting Client can attempt to connect to
the selectedCastingPlayer using
[Matter User Directed Commissioning (UDC)](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/rendezvous/UserDirectedCommissioning.adoc).
The Matter TV Casting library locally caches information required to reconnect

> to a CastingPlayer, once it has commissioned with it. After that, it is able to

Reviewer: ```suggestion
to a CastingPlayer, once it has been commissioned by it. After that, it is able to
```

skip the full UDC process by establishing CASE with the CastingPlayer directly.

> Once connected, the CastingPlayer object would contain the list of available

Reviewer: ```suggestion
Once connected, the CastingPlayer object will contain the list of available
```

> Endpoints on it.

Reviewer: ```suggestion
Endpoints on that CastingPlayer.
```

> On Android, the Casting Client can connect to a `CastingPlayer` by calling
> `connect` on it, and handling any exceptions that may arise in the process.

Reviewer: ```suggestion
On Android, the Casting Client can connect to a `CastingPlayer` by successfully calling
`connect` on it. Remember to handle any exceptions that may arise in the process.
```

```java
// Maximum time in seconds we'll wait for the connection to the CastingPlayer to go through
private static final long CONNECTION_TIMEOUT_SECS = 45;

public static void demoConnection(CastingPlayer castingPlayer) {
    // Connect to the castingPlayer
    CompletableFuture<Void> connectionFuture = castingPlayer.connect(CONNECTION_TIMEOUT_SECS);

    // handle exception, if any

>     connectionFuture.exceptionally(exc -> {

Reviewer 3: You may also want to add the success callback definition for the future here for demonstration purpose.

Author: Done

Product ID, list of supported Clusters, etc).

On Android, for example, it may select an Endpoint with a particular vendorID.

```java
...
Optional<Endpoint> contentAppEndpoint =
    castingPlayer
        .getEndpoints()
        .stream()
        .filter(endpoint -> SAMPLE_CONTENT_APP_VID.equals(endpoint.getVendorId()))
        .findAny();
...
```

On iOS, it can do the above Endpoint selection like the following.

```objectivec
...
if let endpoint: MTREndpoint = castingPlayer.endpoints.filter({ $0.identifier == sampleContentAppVid }).first
...
```

## Interacting with a Casting Endpoint

Once the Casting Client has selected an `Endpoint`, it is ready to
[issue commands](#issuing-commands) to it, [read](#read-operations) current
playback state, and [subscribe](#subscriptions) to playback events.

### Issuing Commands

_{Complete Command invocation examples:
[Android](android/App/app/src/main/java/com/matter/casting/CommandInvocationExample.java)
| [iOS](darwin/TvCasting/TvCasting/MTRCommandInvocationExampleViewModel.swift)}_

To issue a command, the Casting Client can `contentAppEndpoint` on the
CastingPlayer it connected to and handle the command response. It would first

> ensure the endpoint supports the required cluster/command.

Reviewer: This paragraph is not clear. Are you meaning to say the client can get a reference to the `contentAppEndpoint` and then send commands to it?

Author: Yes, thats what I meant. I have tried rephrasing this in the latest version of the diff.

    		contentLauncherResponseFuture.thenApply(
			(response) -> {
				// handle the response
				return null;
			});
}
```

On iOS, the commands can be sent as follows:

```objectivec
...
if(endpoint.hasCluster(MTREndpointClusterTypeContentLauncher))
{
    let cluster: MTRContentLauncherCluster = endpoint.cluster(for: MTREndpointClusterTypeContentLauncher) as! MTRContentLauncherCluster
    if(cluster.canLaunchURL())
    {
        cluster.launch(URL(string: "https://YOUR_URL/id")!) { launchResponse, err in
            self.Log.info("invokeCommand completed with \(err)")
            self.status = "ContentLauncher.LaunchURL completed!"

            // consume the launchResponse
        }
    }
    else
    {
        self.status = "Cannot launchURL on cluster"
    }
}
else
{
    self.status = "No ContentLauncher cluster support found on the endpoint"
}
...
```

### Read Operations

_{Complete Attribute Read examples:
[Android](android/App/app/src/main/java/com/matter/casting/AttributeReadExample.java)
| [iOS](darwin/TvCasting/TvCasting/MTRAttributeReadExampleViewModel.swift)}_

The CastingClient may read an Attribute from the `contentAppEndpoint` on the

> CastingPlayer. It would ensure that the desired cluster / attribute are

Reviewer: ```suggestion
CastingPlayer. It should ensure that the desired cluster / attribute are
```

available for reading on the endpoint before trying to read it.

On Android, an attribute can be read with implementation like the one below.

```java
// check if the Endpoint supports the Cluster to use
if (contentAppEndpoint.get().hasCluster(MediaPlayback.class)) {
    // get the Cluster to use
    MediaPlayback mediaPlayback = contentAppEndpoint.get().getCluster(MediaPlayback.class);

    Optional<Attribute<MediaPlayback.CurrentState>> currentStateAttribute = mediaPlayback.getCurrentState();

>     if(currentStateAttribute.isPresent() && currentStateAttribute.get().isAvailable())

Reviewer 3: Is the need for both isPresent and isAvailable to differentiate between the media playback cluster server not setting the attribute versus explicitly setting it to null? Does the client care in this case? The reason is I'm wondering if Attribute.isAvailable is sufficient to handle both cases.

Author: In the Android APIs we are returning an Optional which is why I am making an isPresent check. Either we should return the Attribute itself instead of an Optional\<Attribute> - if we just want to check for isAvailable, or we check both, or we get rid of the isAvailable method.

Reviewer 3: I think we should return the Attribute always and have the isAvailable() check in it. This way, clients can subscribe to it even if it's not available.

Author: Done. Returning an Attribute instead of the Optional\<Attribute\<>> and the Attribute has an isAvailable method that we are calling here.

    {
        CompletableFuture<MediaPlayback.CurrentState> currentStateValue = currentStateAttribute.get().read();
        try {
            Log.i(TAG, "Read Current State value: " + currentStateValue.get(1, TimeUnit.SECONDS));
        } catch (ExecutionException | InterruptedException | TimeoutException e) {
            Log.e(TAG, "Exception when reading CurrentState " + e);
        }
    }
    else
    {
        Log.e(TAG, "Attribute unavailable on the selected endpoint");
    }
}
else
{
    Log.e(TAG, "Required cluster not found on the selected endpoint");
}
```

On iOS, an attribute can be read with implementation like the one below.

```objectivec
if(endpoint.hasCluster(MTREndpointClusterTypeContentLauncher))
{
    let cluster: MTRMediaPlaybackCluster = endpoint.cluster(for: MTREndpointClusterTypeMediaPlayback) as! MTRMediaPlaybackCluster

>     let currentStateAttribute: MTRAttribute<MTRCurrentState> = cluster.currentState

Reviewer 3: It looks like in the iOS example MTRAttribute handles the optional. Can Android example leverage the same?

Author: I've added an isAvailable method and checked if it returns true here. Is that what you meant?

Reviewer 3: I meant can't MTRAttribute encapsulate the logic of what optionals do today? I.e for each attribute definition, it will always return a MTRAttribute instance but has itself a isAvailable check in it that consumers can use to check if the value is available.

Author: This was already done, right? See line#582 for eg.

> 

Reviewer 3: Should we first add the isAvailable check?

Author: Done

    currentStateAttribute.read { currentStateValue, err in
        if(err == nil)
        {
            self.status = "Read CurrentState value \(String(describing: currentStateValue))"
        }
        else
        {
            self.status = "Error when reading CurrentState value \(String(describing: err))"
        }
    }
}
else
{
    self.status = "No MediaPlayback cluster supporting endpoint found"
}
```

### Subscriptions

_{Complete Attribute subscription examples:
[Android](android/App/app/src/main/java/com/matter/casting/AttributeSubscriptionExample.java)
|
[iOS](darwin/TvCasting/TvCasting/MTRAttributeSubscriptionExampleViewModel.swift)}_

> A Casting Client may subscribe to an attribute from the `contentAppEndpoint` on

Reviewer: ```suggestion
A Casting Client may subscribe to a set of attributes on the `contentAppEndpoint` of
```

> the CastingPlayer to get event notifications related to that attribute.

Reviewer: ```suggestion
the CastingPlayer to get data reports when the attributes change.
```

On Android, an attribute can be subscribed to in the following way.

```java
// check if the Endpoint supports the Cluster to use
if (contentAppEndpoint.get().hasCluster(MediaPlayback.class)) {
    // get the Cluster to use
     MediaPlayback mediaPlayback = contentAppEndpoint.get().getCluster(MediaPlayback.class);

>     Optional<Attribute<MediaPlayback.CurrentState>> currentStateAttribute = mediaPlayback.getCurrentState();

Reviewer 3: Same comment as above

    if(currentStateAttribute.isPresent() && currentStateAttribute.get().isAvailable())
    {
        currentStateAttribute.get().addObserver(new Attribute.Listener() {
            @Override
            public void onError(Error error) {
                Log.e(TAG, "Error when listening to CurrentState " + error);
            }

            @Override

>             public void onChange(Object currentStateValue) {

Reviewer 3: Should we pass in the older state as well?

Author: Done

                Log.i(TAG, "CurrentState changed to value " + currentStateValue);
            }

>         }, 0, 1);

Reviewer 3: Can you replace 0 and 1 here by a readable constant?
```suggestion
}, MIN_INTERVAL, MAX_INTERVAL);
```

Author: Done

# Matter Casting APIs

Matter Casting consists of three parts:

-   **The mobile app**: For most content providers, this would be your

>     consumer-facing mobile app. By making your moile app a Matter "Casting

Author: Fixed in project-chip#29444

    Client", you enable the user to discover casting targets, cast content, and
    control casting sessions. The
    [example Matter tv-casting-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-casting-app)
    for Android / iOS and Linux builds on top of the Matter SDK to demonstrate
    how a TV Casting mobile app works.
-   **The TV content app**: For most content providers, this would be your
    consumer-facing app on a Smart TV. By enhancing your TV app to act as a
    Matter "Content app", you enable Matter Casting Clients to cast content. The
    [example Matter content-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-app/android/App/content-app)
    for Android builds on top of the Matter SDK to demonstrate how a TV Content
    app works.
-   **The TV platform app**: The TV platform app is generally implemented by the

>     TV manufacturer and provides common capabilities around media playback on

Author: Updated in project-chip#29444

samples so you can see the experience end to end.

## Introduction

A Casting Client (e.g. a mobile phone app) is expected to be a Matter
Commissionable Node and a `CastingPlayer` (i.e. a TV) is expected to be a Matter
Commissioner. In the context of the
[Matter Video Player architecture](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/app_clusters/media/VideoPlayerArchitecture.adoc),
a `CastingPlayer` would map to
[Casting "Video" Player](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/app_clusters/media/VideoPlayerArchitecture.adoc#1-introduction).
The `CastingPlayer` is expected to be hosting one or more `Endpoints` (similar
to
[Content Apps](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/app_clusters/media/VideoPlayerArchitecture.adoc#1-introduction)
in the Matter Video Player architecture) that support one or more Matter Media
`Clusters`.

The steps to start a casting session are:

1. [Initialize](#initialize-the-casting-client) the `CastingClient` using the
   Matter SDK.
1. [Discover](#discover-casting-players) `CastingPlayer` devices using Matter
   Commissioner discovery.

> 1. [Connect](#connect-to-a-casting-player) to the `CastingPlayer` to discover
>    available endpoints. By connecting, the 'CastingClient' is first sending a
>    User Directed Commissioning (UDC) request to the 'CastingPlayer' device in
>    order to request Matter commissioning by it. For commissioning, the
>    'CastingPlayer' will obtain the appropriate user consent to allow a
>    connection from this 'CastingClient' and obtain the setup code needed to
>    commission the 'CastingClient'. The setup code will typically come from a
>    corresponding TV content app or be input by the user.

Reviewer 2: ```suggestion
1. [Connect](#connect-to-a-casting-player) to the `CastingPlayer` to discover
   available endpoints. By connecting, the 'CastingClient' will send a
   User Directed Commissioning (UDC) request to the 'CastingPlayer' device in
   order to make a Matter commissioning request. The
   'CastingPlayer' will then obtain the appropriate user consent to allow a
   connection from this 'CastingClient' and obtain the setup code needed to
   commission the 'CastingClient'. The setup code will typically come from a
   corresponding TV content app or be input by the user.
```

1. [Select](#select-an-endpoint-on-the-casting-player) an available `Endpoint`
   hosted by the `CastingPlayer`.

Next, you're ready to:

1. [Issue commands](#issuing-commands) to the `Endpoint`.
1. [Read](#read-operations) endpoint attributes like playback state.
1. [Subscribe](#subscriptions) to playback events.

## Build and Setup

The Casting Client is expected to consume the Matter TV Casting library built

> for its respective platform to consume any of the APIs described in this

Reviewer: ...for its respective platform which implements the APIs described in this document.

Author: Updated in project-chip#29444

document. Refer to the tv-casting-app READMEs for [Android](android) and
[iOS](darwin) to understand how to build and consume each platform's specicific
libraries.

### Initialize the Casting Client

_{Complete Initialization examples:

> [Android](android/App/app/src/main/java/com/matter/casting/InitializationExample.java)

Reviewer 2: this link didn't work for me, can you confirm the path is correct?

Author: Its correct. Here is the file: [https://github.com/sharadb-amazon/connectedhomeip/blob/casting/examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/InitializationExample.java](https://github.com/sharadb-amazon/connectedhomeip/blob/casting/examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/InitializationExample.java)

The link is just not working on the source branch - and that can be ignored. You can see the link works when you go the Initialization section and click on it here: https://github.com/sharadb-amazon/connectedhomeip/blob/casting/examples/tv-casting-app/APIs.md

| [iOS](darwin/TvCasting/TvCasting/TvCastingApp.swift)}_

A Casting Client must first initialize the Matter SDK and define the following

> [`DataProvider`](android/App/app/src/main/jni/com/matter/casting/params/DataProvider.java)

Reviewer 2: this one didn't work either, maybe just preview is broken? Other links work fine though

https://github.com/sharadb-amazon/connectedhomeip/blob/9e60120c433a864fa7af897ed919d47cf4276c64/examples/tv-casting-app/APIs.md

    ```objectivec
    class AppParametersDataSource : NSObject, MTRDataSource
    {
        func castingAppDidReceiveRequestForRotatingDeviceIdUniqueId(_ sender: Any) -> Data {
            return "YOUR_GENERATED_ROTATING_DEVICE_ID".data(using: .utf8)!      // dummy value for demonstration only
        }
        ...
    }
    ```

2.  **Commissioning Data** - This object contains the passcode, discriminator,
    etc which identify the app and are provided to the CastingPlayer during the
    commissioning process. Refer to the Matter specification's
    [Onboarding Payload](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/qr_code/OnboardingPayload.adoc#ref_OnboardingPayload))
    section for details on commissioning data.

    On Android, define a `commissioningDataProvider` that can provide the
    required values to the CastingApp.

    ```java
    private final static DataProvider<CommissioningData> commissioningDataProvider = () -> {
        CommissioningData commissioningData = new CommissioningData();
        commissioningData.setSetupPasscode(20202021);   // dummy values for demonstration only

>         commissioningData.setDiscriminator(3874);

Reviewer: this also is a dummy value for demonstration only

Author: Updated in project-chip#29444

        return commissioningData;
    };
    ```

    On iOS, add a `func commissioningDataProvider` to the
    `AppParametersDataSource` class defined above, that can provide the required
    values to the `MTRCastingApp`.

    ```objectivec
    func castingAppDidReceiveRequestForCommissioningData(_ sender: Any) -> MTRCommissioningData {
        return MTRCommissioningData(
            passcode: 20202021,         // dummy values for demonstration only

>             discriminator: 3874,

Reviewer: dummy value for demonstration only

Author: Updated in project-chip#29444

            spake2pIterationCount: 1000,
            spake2pVerifier: nil,
            spake2pSalt: nil)
    }
    ```

3.  **Device Attestation Credentials** - This object contains the
    `DeviceAttestationCertificate`, `ProductAttestationIntermediateCertificate`,
    etc. and implements a way to sign messages when called upon by the Matter TV
    Casting Library as part of the
    [Device Attestation process](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/device_attestation/Device_Attestation_Specification.adoc)
    during commissoning.

    On Android, define a `dacProvider` to provide the Casting Client's Device
    Attestation Credentials, by implementing a `DataSource`:

    ```java
    private final static DataProvider<DeviceAttestationCredentials> dacProvider = new DataProvider<DeviceAttestationCredentials>() {

>         private static final String kDevelopmentDAC_Cert_FFF1_8001 = "MIIB5z...<snipped>...CXE1M=";     // dummy values for demonstration only

Reviewer: if all 3 are dummy values for demonstration only, then put the comment above the first one. Otherwise, it could be interpreted to mean that only this first variable is dummy value.

Author: Updated in project-chip#29444

>         private static final String kDevelopmentDAC_PrivateKey_FFF1_8001 = "qrYAror...<snipped>...StE+/8=";

Reviewer: Isn't there a way to get the OS to generate and store the private key? Why not provide that in the example code rather than hardcoding it this way, which would never be done in practice

```java
private static final CastingPlayerChangeListener castingPlayerChangeListener = new CastingPlayerChangeListener()
{
    private final String TAG = CastingPlayerChangeListener.class.getSimpleName();

    @Override
    public void onChanged(List<CastingPlayer> castingPlayers) {
        Log.i(TAG, "Discovered changes to " + castingPlayers.size() + " CastingPlayers");
        // consume changes to the provided castingPlayers
    }

    @Override
    public void onAdded(List<CastingPlayer> castingPlayers) {
        Log.i(TAG, "Discovered " + castingPlayers.size() + " CastingPlayers");
        // consume discovered castingPlayers
    }

    @Override
    public void onRemoved(List<CastingPlayer> castingPlayers) {
        Log.i(TAG, "Removed " + castingPlayers.size() + " CastingPlayers");
        // consume castingPlayers removed or lost from the network
    }
};
```

On iOS, implement a `func addDiscoveredCastingPlayers`,
`func removeDiscoveredCastingPlayers` and `func updateDiscoveredCastingPlayers`
which listen to notifications as Casting Players are added, removed, or updated.

```objectivec
@objc
func didAddDiscoveredCastingPlayers(notification: Notification)
{

>     guard let userInfo = notification.userInfo,

Reviewer: the android version prints out when the callback gets invoked. Should we have iOS behave the same way? Seems like if its useful to log in the Android case then it would also be useful in the iOS case.

    MTRCastingPlayerDiscovery.sharedInstance().start()
    ...
}
```

Note: You will need to connect with a Casting Player as described below to see
the list of Endpoints that they support. Refer to the [Connection](#connection)
section for details on how to discover available endpoints supported by a
Casting Player.

### Connect to a Casting Player

_{Complete Connection examples:
[Android](android/App/app/src/main/java/com/matter/casting/ConnectionExample.java)
| [iOS](darwin/TvCasting/TvCasting/MTRConnectionExampleViewModel.swift)}_

Each CastingPlayer object created during [Discovery](#discovery) contains
information such as deviceName, vendorId, productId, etc. which can help the
user pick the right CastingPlayer. A Casting Client can attempt to connect to
the selectedCastingPlayer using
[Matter User Directed Commissioning (UDC)](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/rendezvous/UserDirectedCommissioning.adoc).
The Matter TV Casting library locally caches information required to reconnect

> to a CastingPlayer, once it has been commissioned by it. After that, it is able

Reviewer: ...once the casting client has been commissioned by the CastingPlayer. After that, the casting client is able...

(eg. be clear what "it" was referring to)

Author: Done in project-chip#29870
